### ISSUE TITLE
Apple pay in webview

### ISSUE THREAD

**Issue by MaksymLebedevs** [react-native-webview/react-native-webview#920]
Is it possible to use Apple Pay on pages opened with react-native-webview? As I understand to use it on pages they should be opened in SFSafariViewController, while react-native-webview uses WKWebView. Is it possible to switch to SFSafariViewController?

**Comment by meshi** [comment]
i'm following this issue

**Comment by BilalArxd** [comment]
I am having same issue using stripe with Apple Pay and Google Pay. Any solution or workaround would be appreciated. Thanks

**Comment by ianneub** [comment]
I updated our app to react-native-webview 7.0.5 from react-native's 0.42 version. We found that Apple Pay now works in our webview app! Later on I added an `onMessage` handler to the WebView and now Apple Pay does not work.

I assume this is another issue, but does anyone else using the `onMessage` handler? Maybe that is why you don't get Apple Pay?

**Comment by esauro** [comment]
I'm having the same issue. I updated to react-native-webview 7.4.3 from react-native 0.60. It didn't work. We have an onMessage handler, I removed the onMessage and it's still not working.

I will try the same with a SFViewController, but that won't provide the UX I want to provide

**Comment by shusain93** [comment]
> In iOS 13, webpages loaded in WKWebView can now accept Apple Pay. In order to protect the security of Apple Pay transactions in WKWebView, Apple Pay cannot be used alongside of script injection APIs such as WKUserScript or evaluateJavaScript(_:completionHandler:).
> 
> If these APIs are invoked before a webpage uses Apple Pay, Apple Pay will be disabled. If a webpage uses Apple Pay before evaluateJavaScript(_:completionHandler:) is invoked, the completion handler will be called with a non-nil NSError. These restrictions are reset every time the top frame is navigated.

https://webkit.org/blog/9674/new-webkit-features-in-safari-13/

**Comment by hmolotsi** [comment]
I'm having this same issue and I'm using react-native-webview 8.0.6. Has anyone gotten this to work? What WebView parameters are required?

**Comment by pxr13** [comment]
I got it to work by putting all the Apple Pay processing code on a page in my web app and using that as the `uri` in the `source` prop of the `WebView` component.

You'll need to put the button and Apple Pay session code in the web app for this to work and don't use the `injectedJavascript` prop in the `WebView` component because that will disable Apple Pay (as the post by @shusain93 shows).

**Comment by temitope** [comment]
> I got it to work by putting all the Apple Pay processing code on a page in my web app and using that as the `uri` in the `source` prop of the `WebView` component.
> 
> You'll need to put the button and Apple Pay session code in the web app for this to work and don't use the `injectedJavascript` prop in the `WebView` component because that will disable Apple Pay (as the post by @shusain93 shows).
@pxr13 
can you show a sample of what the uri value used in `WebView` looked like (e.g. how is the Apple Pay processing code integrated with the uri?) 

Looking to have it work on shopify (which already sets up the button to work with apple pay in the Safari browser). thanks

**Comment by kfvsc** [comment]
Problem at 230 string RNCWebView.m:
`[wkWebViewConfig.userContentController addUserScript:script];`
Comment or remove it and Apple Pay be work.

**Comment by martinattin** [comment]
> Problem at 230 string RNCWebView.m:
> `[wkWebViewConfig.userContentController addUserScript:script];`
> Comment or remove it and Apple Pay be work.

Yes it works, but only if we remove onmessage prop from webview

Is there any other way which i can use to get callback?

**Comment by martinattin** [comment]
> I updated our app to react-native-webview 7.4.3 from react-native's 0.42 version. We found that Apple Pay now works in our webview app!
> 
> Later on I added an `onMessage` handler to the WebView and now Apple Pay does not work.
> 
> I assume this is another issue, but is anyone else using the `onMessage` handler? Maybe that is why you don't get Apple Pay?

Is there any other way which i can use along with webview to get callback from  web to app?

**Comment by charkcl** [comment]
> I updated our app to react-native-webview 7.4.3 from react-native's 0.42 version. We found that Apple Pay now works in our webview app!
> 
> Later on I added an `onMessage` handler to the WebView and now Apple Pay does not work.
> 
> I assume this is another issue, but is anyone else using the `onMessage` handler? Maybe that is why you don't get Apple Pay?

@ianneub Thanks for your tips!! Does Google Pay work on Android after the upgrade of react-native-webview?

**Comment by ianneub** [comment]
> @ianneub Thanks for your tips!! Does Google Pay work on Android after the upgrade of react-native-webview?

Sorry @charkcl , I am not using Google Pay.

**Comment by Pyroboomka** [comment]
We also use webview to handle user payments in our app.

The problem is in version 8.0.0 guys fixed issues with history via injecting javascript to shim the code. So for me, Apple Pay widget shows up on 7.6.0, but is not present in anything after 8.0.0.

Sucks, but I don't see this working unless some issues which led to 8.0.0 release are fixed.

**Comment by charkcl** [comment]
Hi @Pyroboomka thanks for clarifying. Just wondering if Google Pay works when you use 7.6.0? Thanks

**Comment by ramisalem** [comment]
Hi , For anyone who still has the issue and dose not get it work 
here is what worked for me 
`<WebView useWebKit={true} source={{url: 'https://www.google.com'}} />`
[ RN docs link ](https://reactnative.dev/blog/2018/08/27/wkwebview)

**Comment by RWOverdijk** [comment]
I must say... This is a real bummer. I use communication between the webview and the app to validate the address and such because stripe doesn't have a decent solution for react native.

And now I have to rebuild my entire payment flow, just for apple pay.

Pisses me off to be honest. Are there other workaround possible? Anything that doesn't involve rebuilding the entire thing? Even if it takes days, I'll take it as rebuilding the whole thing will take long.

**Comment by jamonholmgren** [comment]
This appears to be the problem here:

https://github.com/react-native-community/react-native-webview/compare/v7.6.0...v8.0.0#diff-e2c6890fad2eec1786f059cac2d6d839R187

(Thanks to @pxr13 and @kfvsc for their help on this.)

What would be a good solution here without completely rolling back that change?

Perhaps another prop like `shimHistory` that defaults to true, and you could set it to `false` if it's causing problems?

Other ideas?

**Comment by RWOverdijk** [comment]
I've stripped everything down to a bare minimum and still couldn't get apple pay to show up. It wasn't until I replaced it with a SFSafariViewController popup/window that I got it working.

I know there have been updates to webkit to get this working, but aren't those OS dependent? And as such, not a good idea to use yet given that not everyone will have it yet?

I'm throwing out a lot of assumptions because I don't know the answers. But the TL;DR; is:

1. Will this ever work with a WKWebView?
2. If so, is it OS version dependent?
3. If not, what would I have to do (locally, as a patch) to get it to work?

I don't mind working off of a patch file until I figure out a more permanent solution.

**Comment by temitope** [comment]
Stumped me for months. I had to downgrade to v7.6.0 to get Apple Pay
working.

On Tue, Apr 28, 2020 at 12:48 AM Roberto Wesley Overdijk <
notifications@github.com> wrote:

> I've stripped everything down to a bare minimum and still couldn't get
> apple pay to show up. It wasn't until I replaced it with a
> SFSafariViewController popup/window that I got it working.
>
> I know there have been updates to webkit to get this working, but aren't
> those OS dependent? And as such, not a good idea to use yet given that not
> everyone will have it yet?
>
> I'm throwing out a lot of assumptions because I don't know the answers.
> But the TL;DR; is:
>
>    1. Will this ever work with a WKWebView?
>    2. If so, is it OS version dependent?
>    3. If not, what would I have to do (locally, as a patch) to get it to
>    work?
>
> I don't mind working off of a patch file until I figure out a more
> permanent solution.
>
> —
>
> --
Sent from Mobile

**Comment by RWOverdijk** [comment]
@temitope I am on 7.6.0 and it's not working for me. Anything else that needs to happen?

**Comment by temitope** [comment]
> @temitope I am on 7.6.0 and it's not working for me. Anything else that needs to happen?

🤔 no. I’ve used t he downgrade on multiple projects now with success for Apple Pay. 

You already reinstalled the pods and ensured it updated the version?

I must note that I’m not using any extra props like ja injection.

**Comment by RWOverdijk** [comment]
I have this:

```jsx
      <WebView
        incognito
        source={{ uri: 'https://stripe-payments-demo.appspot.com/' }}
        cacheMode='LOAD_NO_CACHE'
        onNavigationStateChange={handleWebViewNavigationStateChange}
        onError={syntheticEvent => {
          const { nativeEvent } = syntheticEvent
          console.error('WebView error: ', nativeEvent)
        }}
      />
```

But I also tried this:

```jsx
      <WebView
        source={{ uri: 'https://stripe-payments-demo.appspot.com/' }}
      />
```

And the apple pay button doesn't show up. I am on 7.6.0 now, and that link (the stripe demo one) should show an apple pay button but it doesn't. It works in safari.

Do you see something I'm doing wrong there?

**Comment by RWOverdijk** [comment]
@temitope I also tried this with an entirely new project just now and it also doesn't work.

Simply did:

1. `npx react-native init test`, 
2. `yarn add react-native-webview@7.6.0`
3. Update App.js to this:

```js
import React from 'react';
import { WebView } from 'react-native-webview'

const App = () => {
  return  <WebView
    source={{ uri: 'https://stripe-payments-demo.appspot.com/' }}
  />;
};

export default App;
```
4. Setup team in xcode
5. Run on device

What am I missing here?

**Comment by Pyroboomka** [comment]
@RWOverdijk I tested on my setup which works with our payment processor javascript widget, and apple pay button does not show up. So probably problem is not on your end :)

**Comment by RWOverdijk** [comment]
@Pyroboomka So are you saying the stripe link is broken? Just trying to find out where the problem is 😅

**Comment by Ciberusps** [comment]
@RWOverdijk tested on 7.6.0 it works [apple pay demo](https://applepaydemo.apple.com)
but dont work with stripe(link u posted above) 
and also dont work with versions >7.6.0
![image](https://user-images.githubusercontent.com/14001879/80563742-9a2eb800-89f4-11ea-8c8b-a466b4005fcc.png)

**Comment by RWOverdijk** [comment]
@Ciberusps That's a simulator. It always works on a simulator. But to make sure I tried it on a real device and no dice. "Your browser does not support apple pay"

**Comment by temitope** [comment]
> @temitope I also tried this with an entirely new project just now and it also doesn't work.
> 
> 
> 
> Simply did:
> 
> 
> 
> 1. `npx react-native init test`, 
> 
> 2. `yarn add react-native-webview@7.6.0`
> 
> 3. Update App.js to this:
> 
> 
> 
> ```js
> 
> import React from 'react';
> 
> import { WebView } from 'react-native-webview'
> 
> 
> 
> const App = () => {
> 
>   return  <WebView
> 
>     source={{ uri: 'https://stripe-payments-demo.appspot.com/' }}
> 
>   />;
> 
> };
> 
> 
> 
> export default App;
> 
> ```
> 
> 4. Setup team in xcode
> 
> 5. Run on device
> 
> 
> 
> What am I missing here?

@RWoverdijk

Hmm this looks right except I don’t see a step for installing pods. Here. Is that just implied? I explicitly installed the pods and ensured the Podfile.lock reflected the change.

**Comment by temitope** [comment]
This link doesn’t show Apple Pay even on my mobile chrome browser directly. Just card option is visible.  https://stripe-payments-demo.appspot.com/

**Comment by RWOverdijk** [comment]
Because chrome doesn't support apple pay. You have to use Safari (if you have Apple pay set up, otherwise it still won't work).

And yes, sorry, I missed the pod install step 😅

**Comment by RWOverdijk** [comment]
@temitope Because chrome doesn't support apple pay. You have to use Safari (if you have Apple pay set up, otherwise it still won't work).

And yes, sorry, I missed the pod install step 😅

**Comment by temitope** [comment]
> @temitope Because chrome doesn't support apple pay. You have to use Safari (if you have Apple pay set up, otherwise it still won't work).
> 
> 
> 
> And yes, sorry, I missed the pod install step 😅

:) thanks you’re right there. So it still want working for you after the pod install? 🤔

**Comment by RWOverdijk** [comment]
It doesn't work, no. So for now I'm using expo-web-browser.

**Comment by ramisalem** [comment]
Hello guys, I am back with another issue, Apple pay button is disappeared after receiving post message from the webView component, 


I am trying to get the total price via onPost massage from webView I basically run the following in my component 

`<WebView
          style={styles.WebView}
          source={{
            uri: 'https://1cc80937.ngrok.io/apple-pay-js/',
          }}   
           ref={webviewRef}
          onLoad={ () => {
            webviewRef.current.postMessage(40)
          }} 
          javaScriptEnabled={true}
    />` 

receiving massage event in webPage as follows 

 ` window.addEventListener(
  "message",
  function(event) {
    document.getElementById("price").innerText = event.data ;
    var script = document.createElement("script");
    script.src = "./apple.js";
    document.getElementsByTagName("head")[0].appendChild(script);
  },
  false
);
`

after apple.js file is loaded the apple pay button disappeared

**Comment by RWOverdijk** [comment]
Any form of js code (which includes posting messages) will make it not work. It's referenced in this issue multiple times if you want details.

If you even get it to work at all, that's pretty odd imo because I know I can't 😄

**Comment by ramisalem** [comment]
oh my god @RWOverdijk I am gonna cry

**Comment by RWOverdijk** [comment]
I did already. Let it all out. It sucks.

**Comment by ramisalem** [comment]
how is your exprice with   Expo's WebBrowser ?

**Comment by RWOverdijk** [comment]
@ramisalem So far it's working fine. It's just very important that you also have universal/deep links set up.

**Comment by ramisalem** [comment]
@RWOverdijk  Thanks,I will follow the guide

**Comment by ramisalem** [comment]
Hi, I have not find a way I could send paramter to expo-web-browser when opnning the links @RWOverdijk

**Comment by RWOverdijk** [comment]
Because you can't. Unless you put it in the url when initially opening it.

**Comment by ramisalem** [comment]
Ok Thanks

**Comment by savlslipchuk** [comment]
@ramisalem

**Comment by savlslipchuk** [comment]
@ramisalem Hey. I faced the same problem, when using "onMessage" the ApplePay payment button disappears. Have you fixed it?

**Comment by ramisalem** [comment]
> Any form of js code (which includes posting messages) will make it not work. It's referenced in this issue multiple times if you want details.
> 
> If you even get it to work at all, that's pretty odd imo because I know I can't 😄

Do not use onMassage or onPostMsaage

**Comment by savlslipchuk** [comment]
@ramisalem  What did you replace it with onMessage? How do you transfer data from a web page to an application?  I found a rather strange solution, use two components webView. do not judge strictly, it may be useful for someone
<img width="593" alt="Снимок экрана 2020-05-04 в 14 04 40" src="https://user-images.githubusercontent.com/31785918/80960163-005e7500-8e11-11ea-8966-30498aed4429.png">

**Comment by ramisalem** [comment]
@savlslipchuk I basically pass the data to the URL initially when opping it  as params for example 
yourURL?param=etc 

and Instead of onMassage , I use my back-end to tell me if the payment is completed or not 

The app goes to background when the pop-up sheat is opping so I used AppState to check of the user comes from background to foreground , if so I call my back-end 

```  const handleAppStateChange = async newState => {
    if (newState === 'active' ) {
      // ur stuff 
    }
  };

  useEffect(() => {
    AppState.addEventListener('change', handleAppSate);

    return () => {
      AppState.removeEventListener('change', handleAppSate);
    };
  }, []);
 ```

**Comment by RWOverdijk** [comment]
I like that your code gives massages

**Comment by ramisalem** [comment]
what do mean by gives massages ? I do not use any js form in webView

**Comment by RWOverdijk** [comment]
@ramisalem You wrote "massages" instead of "messages". It was a joke. But it didn't land.

**Comment by ramisalem** [comment]
@RWOverdijk lol, landed just now

**Comment by jamonholmgren** [comment]
You always leave me sated.

![image](https://user-images.githubusercontent.com/1479215/81210065-2fe2be00-8f86-11ea-8d47-3221ed076648.png)

**Comment by RWOverdijk** [comment]
Oh look, you do have time! :trollface: Maybe you can fix some issues? 😇

**Comment by jamonholmgren** [comment]
![disappears in a puff of smoke](https://media1.tenor.com/images/ab42c02fe0225129611e6b9b3fb4dc7a/tenor.gif?itemid=5138387)

**Comment by ramisalem** [comment]
https://twitter.com/i/status/1242996244152119296

**Comment by ramisalem** [comment]
![tumblr_oacycl3C6g1s2mmayo1_400](https://user-images.githubusercontent.com/25037017/81289139-82ee6b00-9083-11ea-87e2-4088c8853202.gif)

**Comment by ramisalem** [comment]
I am looking for a way I could find out if user's device supports apple pay without using the webView I would like to show the web view only the device is support apple pay, I know that this question is not related to the opened issue here but it's better then making jokes of my code

**Comment by RWOverdijk** [comment]
@ramisalem That depends on your method of choice. Tipsi does this (when it works). But I'd just assume it works, for example with stripe the button won't even show up if it's not supported.

About the jokes, we were just laughing at typos which in hindsight was a bit rude. Sorry for that.

**Comment by ramisalem** [comment]
@ RWOverdijk I do not use stripe.

About the apple button and the payment process, It is working well in my expo app

**Comment by RWOverdijk** [comment]
I don't know to be honest. I'm also using expo and I couldn't get it to work here.

**Comment by ramisalem** [comment]
strange

**Comment by tdoran** [comment]
> I don't know to be honest. I'm also using expo and I couldn't get it to work here.

Can I ask - have you now got it working with this Expo Web Browser? 
https://docs.expo.io/versions/latest/sdk/webbrowser/

If so, how do you handle communication between the webview and the app now? I can get the apple pay button to show up, but I can't see any way of handling communication between app & webpage using Expo's web browser.

**Comment by RWOverdijk** [comment]
@tdoran I did get it working. And I'll say this once and never repeat again (so if I link anyone to this comment in the future: I'm sorry but I'm not sorry).

I get that it's tempting to want to communicate between the app and the iframe/webview/browser screen... But don't. It's a dumb idea. It's like putting in a screw using a cement mixer. You'll get it to work eventually, maybe, but you'll also have cement everywhere and a bunch of people trying to lock you up for being insane.

And I get to say that, because I tried and I was also dumb for trying. These browser screens simply aren't made to do this work. And you'll safe yourself a lot of headaches if you simply don't even try.

Instead, make it not matter. You çan deep link back to the app when you're done, you can use websockets to listen for events on a common source, you can set up snapshot listeners with firestore (if that's your thing), you can even use polling after the browser window closes itself or the app state changes (because remember! **The user can exit the app and open safari**). There are a lot of possibilities, but communicating between the two without viewing them as entirely separate applications is not going to work in a stable manner.

So, I love you, all of you, silly kids. But don't make the same mistake I made. Un-stuck your mind from the "how to do this" mindset, and embrace the "lol, should I even be trying this?" mindset.

**Comment by YousefAlsbaihi** [comment]
Short answer, 
if you try this demo from apple in webview "https://applepaydemo.apple.com/" it will not work until you disable javascript in the webview like this

```
<WebView
     source={{ uri: "https://applepaydemo.apple.com/" }}
     javaScriptEnable={false} />
```

as for Stripe payment, i tried everything possible even changed the user agent and it didn't show up on the webview, i started to think it's stripe problem not webview.

**Comment by RWOverdijk** [comment]
@YousefAlsbaihi `javaScriptEnabled`*

And Without _any_ javascript the webview becomes pretty useless, which is why stripe doesn't work (it needs JavaScript).

**Comment by Pyroboomka** [comment]
We ended with using onNavigationStateChange prop, with our backend that serves the page doing something like ``` history.pushState(null, null, '/success|failure'); location.reload(); ``` when processing succeeded or failed so provided handler got the event. RN side handler was basically copied from the docs. Works fine for our use case.

**Comment by strdr4605** [comment]
For me, it only works with `v7.6.0` and no `onMessage`

```js
<WebView
  ref={webView}
  source={{ uri: 'https://applepaydemo.apple.com/' }}
  onLoadEnd={onWebViewLoadEnd}
  // onMessage={onWebViewMessage}
/>
```

But I need `onMessage` to handle communication. Any ideas?!

**Comment by losh11** [comment]
Hardcoding to `v7.6.0` will be useless in the future, as I imagine that `react-native-webview` v7 is not being maintained to stay compliant with Apple rules (which are forever-changing). Hopefully a fix is implemented soon.

**Comment by ElegantSoft** [comment]
> Problem at 230 string RNCWebView.m:
> `[wkWebViewConfig.userContentController addUserScript:script];`
> Comment or remove it and Apple Pay be work.

this solved my problem Thanks

**Comment by aviggiano** [comment]
Related to #1242

**Comment by safaiyeh** [comment]
> Problem at 230 string RNCWebView.m:
> `[wkWebViewConfig.userContentController addUserScript:script];`
> Comment or remove it and Apple Pay be work.

Could you submit a PR for this?

**Comment by sajal8759** [comment]
I am using stripe checkout for payment in webview but apple pay button not showing but in safari it will show.  
`<WebView
          source={{
            uri,
          }}
          ref={ref => (this.webview = ref)}
          onNavigationStateChange={this.handleResponse}
          style={{flex: 1, backgroundColor: 'transparent'}}
          onLoad={() => this.setState({loading: false})}
        />`

react-native: 0.61.5 && react-native-webview: ^9.0.1 ...
anyone idea how to fixed..

**Comment by losh11** [comment]
@sajal8759 right now the only fix is to downgrade `react-native-webview` to `v7.6.0`. Make sure to hardcode to that specific version. We're hoping that in the future the maintainers of this repo make `react-native-webview` configurable so we can disable the changes that break Apple Pay support.

**Comment by sajal8759** [comment]
@losh11 if downgrade then some functions not working like onNavigationStateChange .. any other solution?

**Comment by ramisalem** [comment]
> @sajal8759 right now the only fix is to downgrade `react-native-webview` to `v7.6.0`. Make sure to hardcode to that specific version. We're hoping that in the future the maintainers of this repo make `react-native-webview` configurable so we can disable the changes that break Apple Pay support.

I am using expo and it was working well in expo 36  version but when I upgraded expo to expo 38, the webView package updated to 9.4.0 which breaks  apple pay, I have tried to downgrade  react-native-web-view to v7.6.0 but it did not work I am also getting react-native-webview - expected version range: 9.4.0 - actual version installed: 7.6.0 warning message when I start the  server

**Comment by yunuseskisan** [comment]
Attempted all suggestions in the thread above and still unable to get apple pay to work in a web view.

Ive got the following expo snack https://snack.expo.io/@yunus_eskisan/web-view-apple-pay

If anyone is able to get it to work please let us in the thread know with a working solution.

**Comment by Ciberusps** [comment]
most commented issue didn't fixed in 1+ year already 😎😎😎💪💪💪
![image](https://user-images.githubusercontent.com/14001879/95158505-6b3b2000-07a4-11eb-83d2-f59cde086168.png)

**Comment by yunuseskisan** [comment]
For all of you still interested in using apple pay on Stripe or any other gateway you cannot use WebView. The trick is to you a BrowseView instead, and have back end which receives the user redirections and links back into app.

**Comment by Overtorment** [comment]
switched to `react-native-inappbrowser-reborn` for applepay.
for general browsing still use `react-native-webview`

**Comment by Ciberusps** [comment]
I fixed this issue in my app, so I want to resume all info in this thread and tell about pitfalls u can run into.

### tl;dr
**expo** - there is no solution and wont be in near future, cause even if tomorrow its will be fixed by somebody u'll need to wait for months until `react-native-webview` will be updated in Expo SDK(e.g. 39.0.3 uses [`v10.7.0`](https://github.com/expo/expo/blob/master/CHANGELOG.md#-3rd-party-library-updates-1)). The only way to fix this issue - eject as i did.

**react-native**(aka "bare workflow" for expo users):
- What causes issue - script that "injected" by default on webview start - `html5HistoryAPIShimSource`
- How to fix - just comment line in `/node_modules/react-native-webview/apple/RNCWebView.m` like shown below
```lang-m
WKUserScript *script = [[WKUserScript alloc] initWithSource:html5HistoryAPIShimSource injectionTime:WKUserScriptInjectionTimeAtDocumentStart forMainFrameOnly:YES];
// [wkWebViewConfig.userContentController addUserScript:script]; // this line that inject "html5HistoryAPIShimSource" on start
```
in `v10.9.2` line number is `1270`.

### Some details and pitfalls:
Many people in this discussion test apple pay in wrong way and dont know device requirements [as i was](https://github.com/react-native-webview/react-native-webview/issues/920#issuecomment-621001681):
- dont use simulator
- dont use https://applepaydemo.apple.com for tests
- use device with iOS 13+. All devices starting from "iPhone 6s / 6s Plus" support latest iOS, also u can check iOS version compatibility here https://iosref.com/ios. Reason why u should use devices with iOS 13+ cause [Apple Pay support in WKWebView appeared only in Safari 13](https://developer.apple.com/documentation/safari-release-notes/safari-13-release-notes#Payment-Request-API) remind that safari major version = iOS version
- dont use any script injections e.g. `onMessage`, `injectedJavaScript` cause [they use `WKUserScript` for injection in webpage and disables Apple Pay](https://developer.apple.com/documentation/safari-release-notes/safari-13-release-notes#Payment-Request-API). But u'll have question how to communicate between webview and app - if u need apple pay in most cases u are integrating 3rd party checkout(e.g. stripe) -> all u need "close" webview on success or fail -> many checkout providers have "callback urls" e.g. stripe have `success_url`, just use it, redirect user back into your app on success/fail via [deep link](https://reactnative.dev/docs/linking).

### How to fix this issue
Main problem is history shimming, tbh i dont get why it's so important to inject shimming script instantly and why nobody fixed it already or didnt made it optionally, if i'll have time i'll try to make pr to fix this.

Hope it will help somebody
P.S. some info mb incorrect text me and i'll edit

**Comment by NomaanAhmed** [comment]
Any solution expected here? Seems like this is a pretty popular issue and I think many of us would love to see a solution from the package itself instead of using a workaround.

Also, regarding this:
> **expo** - there is no solution and wont be in near future

Would forking this package, modifying `RNCWebView.m`, and using the forked package in expo work? Using `v7.6.0` as suggested has many concerns, so I think forking the latest version and modifying the above file is an alternative, but haven't explored it yet.

**Comment by mav10** [comment]
Hi everyone!

I've just googled that thread after "ages" to explore "how to pay in RN apps".
And as for me, you _(most of you)_ are trying to catch both rabbits by one shot.

Why don't you use external libs for Apple/Google pay scenario?
As for me, WebView is a good thing for payment getaways without SDK (at least for native platforms) and in the case of credit/debit cards.
Why are you trying to use native payment in WebView?

Probably I haven't enough dived deep into it yet. But I just do not understand the real problem.

**Comment by losh11** [comment]
_still searching for solutions_

**Comment by Ciberusps** [comment]
If anybody still waiting for "quick fix" published package with temporary fix(forked v11.2.1) - https://github.com/winstrike/react-native-webview
P.S. Works only with react-native aka "bare workflow", [expo wont work](https://github.com/react-native-webview/react-native-webview/issues/920#issuecomment-720305564)

**Comment by angel-droth-motive** [comment]
For anyone confused on how ton install @Ciberusps fork, you need to add a .npmrc file in your root directory with the following:
```
registry=https://registry.npmjs.org
@winstrike:registry=https://npm.pkg.github.com
//npm.pkg.github.com/:_authToken=XXXX
```

**Comment by netmaxt3r** [comment]
I have created a pull request #1946 if some one can look into it will be nice, this allow new props so conditionally we can applepay enabled webview. If some one using [patch-package](https://www.npmjs.com/package/patch-package) for react-native project use below patch `react-native-webview+11.3.2.patch`
```diff
diff --git a/node_modules/react-native-webview/apple/RNCWebView.h b/node_modules/react-native-webview/apple/RNCWebView.h
index 805031e..1f48fed 100644
--- a/node_modules/react-native-webview/apple/RNCWebView.h
+++ b/node_modules/react-native-webview/apple/RNCWebView.h
@@ -68,6 +68,7 @@
 @property (nonatomic, assign) BOOL ignoreSilentHardwareSwitch;
 @property (nonatomic, copy) NSString * _Nullable allowingReadAccessToURL;
 @property (nonatomic, assign) BOOL pullToRefreshEnabled;
+@property (nonatomic, assign) BOOL applePayCompactMode;
 #if !TARGET_OS_OSX
 @property (nonatomic, weak) UIRefreshControl * _Nullable refreshControl;
 #endif
diff --git a/node_modules/react-native-webview/apple/RNCWebView.m b/node_modules/react-native-webview/apple/RNCWebView.m
index 9658d03..2e6dc95 100644
--- a/node_modules/react-native-webview/apple/RNCWebView.m
+++ b/node_modules/react-native-webview/apple/RNCWebView.m
@@ -146,7 +146,7 @@ static NSDictionary* customCertificatesForHost;
 #if defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && __IPHONE_OS_VERSION_MAX_ALLOWED >= 130000 /* __IPHONE_13_0 */
     _savedAutomaticallyAdjustsScrollIndicatorInsets = NO;
 #endif
-      
+      _applePayCompactMode = NO;
   }
 
 #if !TARGET_OS_OSX
@@ -1292,6 +1292,13 @@ static NSDictionary* customCertificatesForHost;
 - (void)resetupScripts:(WKWebViewConfiguration *)wkWebViewConfig {
   [wkWebViewConfig.userContentController removeAllUserScripts];
   [wkWebViewConfig.userContentController removeScriptMessageHandlerForName:MessageHandlerName];
+  if( self.applePayCompactMode){
+    if (self.postMessageScript){
+      [wkWebViewConfig.userContentController addScriptMessageHandler:[[RNCWeakScriptMessageDelegate alloc] initWithDelegate:self]
+                                                                       name:MessageHandlerName];
+    }
+    return;
+  }
 
   NSString *html5HistoryAPIShimSource = [NSString stringWithFormat:
     @"(function(history) {\n"
diff --git a/node_modules/react-native-webview/apple/RNCWebViewManager.m b/node_modules/react-native-webview/apple/RNCWebViewManager.m
index 270d3e4..9055961 100644
--- a/node_modules/react-native-webview/apple/RNCWebViewManager.m
+++ b/node_modules/react-native-webview/apple/RNCWebViewManager.m
@@ -99,6 +99,7 @@ RCT_EXPORT_VIEW_PROPERTY(limitsNavigationsToAppBoundDomains, BOOL)
 RCT_EXPORT_VIEW_PROPERTY(messagingEnabled, BOOL)
 RCT_EXPORT_VIEW_PROPERTY(onMessage, RCTDirectEventBlock)
 RCT_EXPORT_VIEW_PROPERTY(onScroll, RCTDirectEventBlock)
+RCT_EXPORT_VIEW_PROPERTY(applePayCompactMode, BOOL)
 
 RCT_EXPORT_METHOD(postMessage:(nonnull NSNumber *)reactTag message:(NSString *)message)
 {
diff --git a/node_modules/react-native-webview/lib/WebViewTypes.d.ts b/node_modules/react-native-webview/lib/WebViewTypes.d.ts
index 354b687..2069620 100644
--- a/node_modules/react-native-webview/lib/WebViewTypes.d.ts
+++ b/node_modules/react-native-webview/lib/WebViewTypes.d.ts
@@ -501,6 +501,19 @@ export interface IOSWebViewProps extends WebViewSharedProps {
      * @platform ios
      */
     limitsNavigationsToAppBoundDomains?: boolean;
+	/**
+	 * A Boolean value which, when set to `true`, WKWebView will be in ApplePay compact mode.
+	 *  Once set, websites will be able to invoke apple pay from React Native Webview.
+	 *  This comes with a cost features like script injection,html5 History , shared Cookies,
+	 *  atEndScript , atStartScript will not work
+	 * {@link https://developer.apple.com/documentation/safari-release-notes/safari-13-release-notes#Payment-Request-API ApplePay Doc}
+	 * if you require to send message to App , webpage has to explicitly call webkit message handler
+	 * @example
+	 *     window.webkit.messageHandlers.ReactNativeWebView.postMessage("hello apple pay")
+	 * from webpage are receive it on onMessage handler on react native side
+	 * @platform ios
+	 */
+    applePayCompactMode?: boolean;
 }
 export interface MacOSWebViewProps extends WebViewSharedProps {
     /**

```

**Comment by losh11** [comment]
*still searching for solutions*

**Comment by arvenjadeaguilar** [comment]
No solutions yet

**Comment by losh11** [comment]
Still searching for solutions

**Comment by gregavola** [comment]
Does anyone know if this works with Google Pay?

**Comment by Ciberusps** [comment]
@gregavola in my app google pay didn't work even with this solution, mb i forget some details but when i tried to pay via google pay in webview it redirected on google page in external browser with some error and we didn't found how to fix this, so we decided just to implement payments via native sdks finally(

**Comment by losh11** [comment]
> ```
> <WebView
>    enableApplePay={true}
>    source={{uri: "https://applepaydemo.apple.com/"}} 
>     automaticallyAdjustContentInsets={false}>
> ```

I'm happy with this issue being closed thanks to this addition.

**Comment by romualdbrisson** [comment]
`<WebView
   enableApplePay={true}
   source={{uri: "https://applepaydemo.apple.com/"}} `
   
This code does not work, neither in a device nor in the simulator.
The message displays: "Your browser does not support Apple Pay on the web".
I am using version 11.14.2 of react native webview.

**Comment by sofian-sunny** [comment]
I am using 
enableApplePay={true}
useWebKit={true}

and "react-native-webview": "^11.14.2"

Its working for me

**Comment by Atom1er** [comment]
I tried all of the above but still not showing up.

**Comment by saulamsal** [comment]
@Atom1er  try enableApplePay={true}

**Comment by AliRehman7141** [comment]
> I am using enableApplePay={true} useWebKit={true}
> 
> and "react-native-webview": "^11.14.2"
> 
> Its working for me

I tried your suggestion but still getting
Cannot run javascript when apple pay is enabled

**Comment by Ciberusps** [comment]
> > I am using enableApplePay={true} useWebKit={true}
> > and "react-native-webview": "^11.14.2"
> > Its working for me
> 
> I tried your suggestion but still getting Cannot run javascript when apple pay is enabled

Its intended, scripts injection in webpages disables ApplePay, so u cant inject scripts, use deep links like described here(last bullet point) https://github.com/react-native-webview/react-native-webview/issues/920#issuecomment-720305564

**Comment by nikitph** [comment]
as we cannot use injected js, the following approach was taken by me for stripe 

`<WebView
          source={{ uri: url }}
          style={{ flex: 1, marginTop: -50, overflow: 'hidden' }}
          enableApplePay={true}
          useWebKit={true}
          automaticallyAdjustContentInsets={false}
          onNavigationStateChange={y => {
            onSuccess(y.url);
          }}
        />`

and a dirty check inside onsuccess like this . Hope this helps 

` const onSuccess = (message: string) => {
    if (message.includes('checkout_success')) {
      setEnable(true);`

**Comment by urwa786** [comment]
@arvenjadeaguilar Thanx for the solution ... it works 
Link to documentation :
[ApplePayIOS](https://github.com/react-native-webview/react-native-webview/blob/master/example/examples/ApplePay.tsx)
